plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTerms.hs

```diff
@@ -0,0 +1,800 @@
+                       n          <- [0..typeArity a],
+                       Just insts <- [findInstantiation ctx n ty a],
+                       x `Set.notMember` fvArgs a
+                     ] of
+                  [] -> mzero
+                  gs -> head gs
+          _ -> mzero
+
+    -- Try to inhabit a constructor `con` of type `conTy` in datatype `d` at type `ty`
+    tryCon d ty (con, conTy)
+      | Set.member d (fvArgs conTy) = mzero   -- <- This is ok, since no mutual recursion
+      | otherwise = do
+          tyctx <- lift $ asks geTypes
+          insts <- maybe mzero pure $ findInstantiation tyctx (typeArity conTy) ty conTy
+          let go tm [] = return tm
+              go tm (InstApp ty : insts) = go (TyInst () tm ty) insts
+              go tm (InstArg ty : insts) = do
+                arg <- findTm ty
+                go (Apply () tm arg) insts
+          go (Var () con) insts
+
+    -- CODE REVIEW: wouldn't it be neat if this existed somewhere?
+    viewApp args (TyApp _ f x) = viewApp (x : args) f
+    viewApp args ty            = (ty, args)
+
+    -- Get the free variables that appear in arguments of a mixed arrow-forall type
+    fvArgs (TyForall _ x _ b) = Set.delete x (fvArgs b)
+    fvArgs (TyFun _ a b)      = ftvTy a <> fvArgs b
+    fvArgs _                  = mempty
+
+-- CODE REVIEW: does this exist anywhere?
+typeArity :: Num a => Type tyname uni ann -> a
+typeArity (TyForall _ _ _ a) = typeArity a
+typeArity (TyFun _ _ b)      = 1 + typeArity b
+typeArity _                  = 0
+
+-- | Generate as small a term as possible to match a given type.
+genAtomicTerm :: Type TyName DefaultUni () -> GenTm (Term TyName Name DefaultUni DefaultFun ())
+genAtomicTerm ty = do
+  ctx  <- asks geTypes
+  vars <- asks geTerms
+  -- First try cheap unification
+  let unifyVar (x, xty) = findInstantiation ctx 0 ty xty
+                       <&> \ tys -> foldl (TyInst ()) (Var () x) [t | InstApp t <- tys]
+  case catMaybes $ map unifyVar $ Map.toList vars of
+    -- If unification didn't work try the heavy-handed `inhabitType`.
+    -- NOTE: We could probably just replace this whole function with
+    -- `inhabitType` and the generators would run fine, but this method
+    -- is probably faster a lot of the time and doesn't rely on the
+    -- order that thins are chosen `inhabitType`. It is also going to generate
+    -- a more even distribution than `inhabitType` (which for performance reasons
+    -- always returns the first thing it finds).
+    [] -> inhabitType ty
+    gs -> liftGen $ elements gs
+
+-- | Generate a term of a given type
+genTermOfType :: Type TyName DefaultUni ()
+              -> GenTm (Term TyName Name DefaultUni DefaultFun ())
+genTermOfType ty = snd <$> genTerm (Just ty)
+
+-- | Generate a term, if the first argument is Nothing then we get something of any type
+-- and if the first argument is `Just ty` we get something of type `ty`.
+genTerm :: Maybe (Type TyName DefaultUni ())
```

Comment:
Is it guaranteed that `genTerm` receives a type of kind `*`? Needs a Note.

Reply:
I've added a comment documenting this requirement. We could also make `genTerm` crash if that's not the case - but that seems a bit over the top.

Reply:
Thanks!